sys/kcm.txt

```diff
@@ -0,0 +1,29 @@
+# Copyright 2015 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <linux/kcm.h>
+
+socket$kcm(domain const[AF_KCM], type flags[kcm_socket_type], proto const[KCMPROTO_CONNECTED]) fd[kcm]
+setsockopt$KCM_RECV_DISABLE(fd fd[kcm], level const[SOL_KCM], opt const[KCM_RECV_DISABLE], val ptr[in, int32], len len[val])
+getsockopt$KCM_RECV_DISABLE(fd fd[kcm], level const[SOL_KCM], opt const[KCM_RECV_DISABLE], val ptr[out, int32], len len[val])
+sendmsg$kcm(fd fd[kcm], msg ptr[in, send_msghdr], f flags[send_flags])
+recvmsg$kcm(fd fd[kcm], msg ptr[in, recv_msghdr], f flags[recv_flags])
+
+ioctl$SIOCKCMATTACH(fd fd[kcm], cmd const[SIOCKCMATTACH], arg ptr[in, kcm_attach])
+ioctl$SIOCKCMUNATTACH(fd fd[kcm], cmd const[SIOCKCMUNATTACH], arg ptr[in, kcm_unattach])
+ioctl$SIOCKCMCLONE(fd fd[kcm], cmd const[SIOCKCMCLONE], arg ptr[inout, kcm_clone])
+
+kcm_socket_type = SOCK_DGRAM, SOCK_SEQPACKET
+
+kcm_attach {
+	fd	fd[sock]
+	bpf_fd	int32
+}
+
+kcm_unattach {
+	fd	fd[sock]
+}
+
+kcm_clone {
+	fd	int32
```

Comment:
kcm_clone ioctl use copy_from_user as the following:

2025         case SIOCKCMCLONE: {
2026                 struct kcm_clone info;
2027                 struct socket *newsock = NULL;
2028 
2029                 if (copy_from_user(&info, (void __user *)arg, sizeof(info)))
2030                         err = -EFAULT;
2031 
2032                 err = kcm_clone(sock, &info, &newsock);
2033 
2034                 if (!err) {
2035                         if (copy_to_user((void __user *)arg, &info,
2036                                          sizeof(info))) {
2037                                 err = -EFAULT;
2038                                 sock_release(newsock);
2039                         }
2040                 }

but in the kcm_clone function, it does not use anything about the info from user.
http://lxr.free-electrons.com/source/net/kcm/kcmsock.c#L1944
 
The fd in the kcm_clone structure is only the output. The copy_from_user is redundant. But I agree with you that fd[sock] is better.